### ISSUE TITLE
new lint: `endless_iter`

### ISSUE THREAD

**Issue by TennyZhuang** [rust-lang/rust-clippy#9699]
### What it does

1. We can create an endless Iterator easily by calling the `Iterator::cycle()` method or by Range literal construction without an upper bound.
  There is a corner case; the `cycle` call on an empty iterator will not create an endless iterator.

3. The endless property will be passed by many methods without an early return branch, e.g.:
  * map
  * filter
  * skip
  * skip_while
  * map_while
  * zip_longest
  * ...
4. We also have some finalizers that will not early return:
  * min
  * max
  * last
  * count
  * for_each
  * reduce
  * collect *
  * sum *
  * product *
  * ...

  There are also some false negative cases because `FromIter,` `Sum` and `Product` are customizable, so there may be some early-return implementation.

Another acceptable name for the lint is `infinite_iter`, both are good.

### Example

```rust
<code>
```

Could be written as:

```rust
<code>
```

**Comment by TennyZhuang** [comment]
> We already have infinite_iter, but I'm unsure if `.cycle()` is caught.

Oh, I check the wrong keyword.

I guess cycle is not covered, I’ll take a look!

**Comment by llogiq** [comment]
Perhaps `maybe-infinite-iter` covers it, because it's not always clear whether an iterator is non-empty.
